### ISSUE TITLE
Using your project

### ISSUE THREAD

**Issue by stereobooster** [peterbe/minimalcss#15]
I tried your project in [react-snap](https://github.com/stereobooster/react-snap), this is small tool to prerender [c-r-a projects](https://github.com/facebookincubator/create-react-app) (and any other SPA). It has much lighter API then penthouse (also penthouse uses puppeteer starting from v1). It is nice, but still removes to much css. I would say this kind of tools work good to remove burden of frameworks (like bootstrap), but pretty hard to use with application css. For application css it would be better to use some kind of [css-in-js solution](https://github.com/MicheleBertoli/css-in-js).

Thanks for the project :wave:

**Comment by peterbe** [comment]
What do you mean it "removes too much css"? Sounds like a bug if anything.

**Comment by peterbe** [comment]
I think there's confusion because I haven't been clear about this. minimalcss will ONLY look at what CSS is needed to complete the first onload. I.e. [after the page has fully loaded](https://github.com/peterbe/minimalcss/blob/d3c1ab602a674b8a7885fdcd72454348cccc27bc/src/run.js#L67). 
So if you have a piece of JavaScript that loads late and changes the DOM then that new DOM won't be what minimalcss uses to compute the minimum CSS. 

However, minimalcss currently doesn't have any opinions or tools to help you use the minimal CSS. Ideally you should include the minimal CSS in your `public/index.html` and also in that `public/index.html` a script that injects the full CSS (i.e. inserting a `<link rel="stylesheet" href="/fullmonty.css">`) *before* the JavaScript starts to load. At least you should *start* it before the JavaScript starts changing the DOM. 

It's early days. These things are tricky to get right. Knowing when the page has fully loaded is a non-trivial thing.

In summary this is what should ideally happen:
1) Load the HTML with the minimal CSS in inline `<style>` block(s)
2) Load the rest of the CSS with JavaScript by changing the DOM to have a new `<link>` node
3) Execute the rest of the React app that potentially changes the DOM further. 

In [my blog post](https://www.peterbe.com/plog/ultrafast-loading-of-css) the site I used was a "regular site" that didn't have any other JavaScript after `onload`. Perhaps too easy as a target?

**Comment by stereobooster** [comment]
This is what I basically did. Except that css is loaded using `<link>` in HTML, it is not generated by JS. But probably I need to try this, and see if there is win in load time.

Also had this problem in my react-snap, when I catch js errors from minified js and it is unreadable. I need some kind of tool to decode this stack trace with sourcemaps. You probably want similar tool in your project. Will let you know if I came out with something.

You also, probably, want to catch error from headless chrome

```js
page.on("console", msg => console.log(`${route}: ${msg}`));
page.on("error", msg => console.log(`${route}: ${msg}`));
page.on("pageerror", msg => console.log(`${route}: ${msg}`));
page.on("requestfailed", msg => console.log(`${route}: ${msg}`));
```

It seems like there are some tools required accompany puppeteer.

**Comment by stereobooster** [comment]
> Also had this problem in my react-snap, when I catch js errors from minified js and it is unreadable

This is what I got https://github.com/stereobooster/sourcemapped-stacktrace-node#example. It is highly experimental, but works in my case

**Comment by stereobooster** [comment]
As of async css loading I found those two resources to be useful:
- https://github.com/filamentgroup/loadCSS
- https://codepen.io/tigt/post/async-css-without-javascript

**Comment by peterbe** [comment]
That [loadCSS](https://github.com/filamentgroup/loadCSS) was what I used in https://www.peterbe.com/plog/ultrafast-loading-of-css
